### PR TITLE
Implemented core list functionality & helper functions

### PR DESCRIPTION
**Pull request by CinchBlue**
Under `src/base/core/` I have implemented basic functions in `lists.hpp` for list manipulation and creation:

- cons
- car
- cdr
- set_car
- set_cdr
- list
- List predicates (proper, improper, null)

I have also included a test case file, `unit-core-lists.cpp`.

**Please look at the test case and play around with the functions. If you think they're all right, please approve it!**

It's missing `length` though. I should probably do it.

### CODE REVIEW

File: src/shaka_scheme/system/core/lists.hpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  // If the last item is not null list, it is not a proper list.
+  return !is_null_list(it);
+}
+
+std::size_t length(NodePtr node) {
+  if (node->get_type() != shaka::Data::Type::DATA_PAIR) {
+    throw shaka::TypeException(1000, "length(): data is not a pair");
+  }
+  int count = 0;
+  for (auto it = core::car(node); !core::is_null_list(it); it = core::cdr(it)) {
```
Comment (btwooton) [inline src/shaka_scheme/system/core/lists.hpp]: Length method looks good, save for one potential flaw. In the for loop, should it be initialized to refer to `core::car(node)` or to node itself?

**Review by btwooton — APPROVED** [review]
I approve this pull request, pending correction of error in the `length(NodePtr node)` method.

File: src/shaka_scheme/system/core/lists.hpp (hunk 1 above)
Comment (CinchBlue) [reply src/shaka_scheme/system/core/lists.hpp]: You're right. I've made a commit to create a test case for length. It's fixed now.

**Comment by CinchBlue** [comment]
I've fixed length. I believe it works. Rebasing...

**Pull request merged**